pinot-segment-local/src/main/java/org/apache/pinot/segment/local/dedup/PartitionDedupMetadataManager.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.pinot.segment.local.utils.RecordInfo;
+import org.apache.pinot.segment.local.utils.tablestate.TableState;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.spi.config.table.HashFunction;
+import org.apache.pinot.spi.data.readers.PrimaryKey;
+import org.apache.pinot.spi.utils.ByteArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class PartitionDedupMetadataManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PartitionDedupMetadataManager.class);
+
+  private final HelixManager _helixManager;
+  private final String _tableNameWithType;
+  private final List<String> _primaryKeyColumns;
+  private final int _partitionId;
+  private final ServerMetrics _serverMetrics;
+  private final HashFunction _hashFunction;
+  private boolean _allSegmentsLoaded;
+
+  // TODO(saurabh) : We can replace this with a ocncurrent Set
+  @VisibleForTesting
+  final ConcurrentHashMap<Object, IndexSegment> _primaryKeySet = new ConcurrentHashMap<>();
+
+  public PartitionDedupMetadataManager(HelixManager helixManager, String tableNameWithType,
+      List<String> primaryKeyColumns, int partitionId, ServerMetrics serverMetrics, HashFunction hashFunction) {
+    _helixManager = helixManager;
+    _tableNameWithType = tableNameWithType;
+    _primaryKeyColumns = primaryKeyColumns;
+    _partitionId = partitionId;
+    _serverMetrics = serverMetrics;
+    _hashFunction = hashFunction;
+  }
+
+  public void addSegment(IndexSegment segment) {
+    // Add all PKs to _primaryKeySet
+    Iterator<RecordInfo> recordInfoIterator = getRecordInfoIterator(segment, _primaryKeyColumns);
+    while (recordInfoIterator.hasNext()) {
+      RecordInfo recordInfo = recordInfoIterator.next();
+      _primaryKeySet.put(HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction), segment);
+    }
+    _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, _partitionId, ServerGauge.DEDUP_PRIMARY_KEYS_COUNT,
+        _primaryKeySet.size());
+  }
+
+  public void removeSegment(IndexSegment segment) {
+    // TODO(saurabh): Explain reload scenario here
+    Iterator<RecordInfo> recordInfoIterator = getRecordInfoIterator(segment, _primaryKeyColumns);
+    while (recordInfoIterator.hasNext()) {
+      RecordInfo recordInfo = recordInfoIterator.next();
+      _primaryKeySet.compute(HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction),
+          (primaryKey, currentSegment) -> {
+            if (currentSegment == segment) {
+              return null;
+            } else {
+              return currentSegment;
+            }
+          });
+    }
+    _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, _partitionId, ServerGauge.DEDUP_PRIMARY_KEYS_COUNT,
+        _primaryKeySet.size());
+  }
+
+  @VisibleForTesting
+  public static Iterator<RecordInfo> getRecordInfoIterator(IndexSegment segment, List<String> primaryKeyColumns) {
```

Comment:
Suggest returning an iterator of `PrimaryKey`. For dedup, we don't need the docId and comparisonValue information from the `RecordInfo`. Similar for the `checkRecordPresentOrUpdate()` which can just take the `PrimaryKey` object. This is not a blocker, so maybe put a TODO and address it later

Reply:
Ack. Didn't see any big impact of changing the method signature to accept PK, hence made that change too.